src/crimson/os/seastore/object_data_handler.cc

```diff
@@ -1083,309 +115,980 @@ ObjectDataHandler::write_ret ObjectDataHandler::prepare_data_reservation(
+    extent_len_t>(overwrite_range.aligned_begin);
+  return ctx.tm.remove_mappings_in_range(
+    ctx.t, overwrite_range.aligned_begin,
+    unaligned_len, std::move(first_mapping));
+}
 
-ObjectDataHandler::write_ret ObjectDataHandler::overwrite(
+// The last step in the multi-mapping-hole-punching scenario: remap
+// the right mapping if it crosses the right edge of the hole's range
+ObjectDataHandler::base_iertr::future<LBAMapping>
+ObjectDataHandler::punch_right_mapping(
   context_t ctx,
-  laddr_t data_base,
-  objaddr_t offset,
-  extent_len_t len,
-  std::optional<bufferlist> &&bl,
-  lba_mapping_list_t &&_pins)
+  overwrite_range_t &overwrite_range,
+  data_t &overwrite_data,
+  LBAMapping right_mapping,
+  op_type_t op_type)
 {
-  if (bl.has_value()) {
-    assert(bl->length() == len);
+  if (right_mapping.is_end() ||
+      overwrite_range.aligned_end <= right_mapping.get_key()) {
+    return ObjectDataHandler::base_iertr::make_ready_future<
+      LBAMapping>(std::move(right_mapping));
   }
-  overwrite_plan_t overwrite_plan(data_base, offset, len, _pins, ctx.tm.get_block_size());
-  return seastar::do_with(
-    std::move(_pins),
-    extent_to_write_list_t(),
-    [ctx, data_base, len, offset, overwrite_plan, bl=std::move(bl), this]
-    (auto &pins, auto &to_write) mutable
-  {
-    LOG_PREFIX(ObjectDataHandler::overwrite);
-    DEBUGT("overwrite: 0x{:x}~0x{:x}",
-           ctx.t,
-           offset,
-           len);
-    ceph_assert(pins.size() >= 1);
-    DEBUGT("overwrite: split overwrite_plan {}", ctx.t, overwrite_plan);
+  return punch_mapping_on_edge(
+    ctx, overwrite_range, overwrite_data,
+    std::move(right_mapping), edge_t::RIGHT, op_type);
+}
 
-    return operate_left(
-      ctx,
-      pins.front(),
-      overwrite_plan
-    ).si_then([ctx, data_base, len, offset, overwrite_plan, bl=std::move(bl),
-               &to_write, &pins, this](auto p) mutable {
-      auto &[left_extent, headbl] = p;
-      if (left_extent) {
-        ceph_assert(left_extent->addr == overwrite_plan.pin_begin);
-        append_extent_to_write(to_write, std::move(*left_extent));
+// punch the hole whose range is within a single pending mapping
+ObjectDataHandler::base_iertr::future<LBAMapping>
+ObjectDataHandler::punch_hole_in_pending_mapping(
+  context_t ctx,
+  overwrite_range_t &overwrite_range,
+  data_t &data,
+  LBAMapping mapping)
+{
+  return merge_pending_edge(ctx, overwrite_range, data, mapping, edge_t::BOTH
+  ).si_then([ctx, mapping=std::move(mapping)]() mutable {
+    return ctx.tm.remove(ctx.t, std::move(mapping));
+  }).handle_error_interruptible(
+    ObjectDataHandler::base_iertr::pass_further{},
+    crimson::ct_error::assert_all{"impossible"}
+  );
+}
+
+ObjectDataHandler::base_iertr::future<LBAMapping>
+ObjectDataHandler::punch_single_mapping_hole(
+  context_t ctx,
+  overwrite_range_t &overwrite_range,
+  data_t &overwrite_data,
+  LBAMapping mapping)
+{
+  if (mapping.is_pending()) {
```

Comment:
Why not reuse the central policy decision maker `get_edge_handle_policy()` ?

Otherwise it is not easy to maintain (adjust/unify) the decisions because they are not in the same place.

Reply:
Yeah, will do

Reply:
Sorry, this method should have been removed, seems I forgot it. Will correct it:-)